File: app/lib/models/syllabus.dart
class Syllabus {
  const Syllabus({
    required this.syllabusId,
    required this.subjectCode,
    required this.subjectName,
    required this.syllabusName,
    required this.credits,
    required this.degreeLevel,
    required this.timeAllocation,
    required this.prerequisite,
    required this.decision,
    required this.description,
    required this.sourcePath,
  });

  factory Syllabus.fromJson(Map<String, dynamic> json) {
    return Syllabus(
      syllabusId: json['syllabus_id'] as int? ?? 0,
      subjectCode: json['subject_code'] as String? ?? '',
      subjectName: json['subject_name'] as String? ?? '',
      syllabusName: json['syllabus_name'] as String? ?? '',
      credits: json['credits']?.toString() ?? '',
      degreeLevel: json['degree_level']?.toString() ?? '',
      timeAllocation: json['time_allocation']?.toString() ?? '',
      prerequisite: json['prerequisite']?.toString() ?? '',
      decision: json['decision'] as String? ?? '',
      description: json['description'] as String? ?? '',
      sourcePath: json['source_path'] as String? ?? '',
    );
  }

  final int syllabusId;
  final String subjectCode;
  final String subjectName;
  final String syllabusName;
  final String credits;
  final String degreeLevel;
  final String timeAllocation;
  final String prerequisite;
  final String decision;
  final String description;
  final String sourcePath;
}

File: crawler/src/models/syllabus.py
from __future__ import annotations

from dataclasses import asdict, dataclass, field
from typing import Any


@dataclass(frozen=True)
class SyllabusIndexItem:
    syllabus_id: int
    subject_code: str
    subject_name: str
    syllabus_name: str
    is_active: bool | None
    is_approved: bool | None
    decision: str
    detail_url: str

    @property
    def canonical_id(self) -> str:
        return f"syllabus:{self.syllabus_id}"

    def to_dict(self) -> dict[str, Any]:
        data = asdict(self)
        data["id"] = self.canonical_id
        return data


@dataclass
class NormalizedSyllabus:
    syllabus_id: int
    subject_code: str
    subject_name: str = ""
    syllabus_name: str = ""
    credits: int | None = None
    degree_level: str = ""
    time_allocation: str = ""
    prerequisite: str = ""
    general_information: dict[str, Any] = field(default_factory=dict)
    active: bool | None = None
    approved: bool | None = None
    decision: str = ""
    description: str = ""
    learning_outcomes: list[dict[str, Any]] = field(default_factory=list)
    assessments: list[dict[str, Any]] = field(default_factory=list)
    schedule: list[dict[str, Any]] = field(default_factory=list)
    constructive_questions: list[dict[str, Any]] = field(default_factory=list)
    materials: list[dict[str, Any]] = field(default_factory=list)
    source_path: str = ""

    @property
    def id(self) -> str:
        return f"syllabus:{self.syllabus_id}"

    def to_dict(self) -> dict[str, Any]:
        data = asdict(self)
        data["id"] = self.id
        return data
